### PR TITLE
derivative for wigner, example of equivariance check, increase generality

### PR DESCRIPTION
**Pull request by bkmi**
Hopefully you like it!

### CODE REVIEW

**Comment by mariogeiger** [comment]
Isn't `ph.derivative_rot_mat` returning 3 matrices? The derivative wrt alpha, wrt beta and wrt gamma?
if so, don't you also want to return 3 tensors?

**Comment by bkmi** [comment]
I updated it to return a 3-tuple of tensors: dDda, dDdb, dDdc.
Does that solve it?

**Comment by bkmi** [comment]
There is also the equivariance check. Perhaps we should wait to merge until we update the equivariance check readme.

**Comment by bkmi** [comment]
I added the equivariance check for the rotation case. Still need to do the parity one.

Note I changed the multiplication file because when using float32, the precision is too low to satisfy the `error < 1e-10` assertion.

**Comment by mariogeiger** [comment]
Thanks Ben, shall I wait the parity test before merge?

**Comment by bkmi** [comment]
Added a fix for #26.
Changed how spherical_harmonics_xyz is calculated.

**Comment by bkmi** [comment]
> Thanks Ben, shall I wait the parity test before merge?

Hmm not sure. I am feeling pretty lazy about writing that right now, but I think there are other good fixes in here. I should have separated them.

**Comment by mariogeiger** [comment]
The `.cpu()` is not necessary.
About the `.detach()` I don’t think we should put it here. 
It’s an informative error for the user that the gradient cannot be computer. 
Moreover if the user changes the default sh by a backwardable one then we don’t want the `.detach()`. 
As a user you can do what you did by doing `sh=lambda x: sh(x.detach())` as argument of Kernel.

**Comment by bkmi** [comment]
Alright. I vote that it's ready if you think it is. Please accept the pull or let me know what I should change.

**Review by mariogeiger — CHANGES_REQUESTED** [review]
for me it looks like `examples/point/equivariance.py` should be put into the test files and use the same format as the other files in this directory.

File: se3cnn/point/kernel.py [hunk 1]
```diff
@@ -113,7 +113,10 @@ def forward(self, r):
         kernel = r.new_zeros(batch, self.n_out, self.n_in)
 
         # precompute all needed spherical harmonics
-        Ys = self.sh(self.set_of_l_filters, r)  # [l_filter * m_filter, batch]
+        if r.requires_grad:
+            raise TypeError("The r input of kernel cannot require a gradient.")
```
Comment (mariogeiger) [inline se3cnn/point/kernel.py]: You cannot do that. Someone might want to allow `r` to require grad if he uses its own `sh` function.
Why do you want to do that? Is it the current error message that is not clear enough?

Comment (bkmi) [reply se3cnn/point/kernel.py]: It was for clarity, but I'm okay with leaving it how it is:

```python
Traceback (most recent call last):
  File "/Users/user/miniconda3/lib/python3.7/site-packages/IPython/core/interactiveshell.py", line 3326, in run_code
    exec(code_obj, self.user_global_ns, self.user_ns)
  File "<ipython-input-2-c6f6414c7b04>", line 1, in <module>
    runfile('/Users/user/sci/cgae/otp_equi.py', wdir='/Users/user/sci/cgae')
  File "/Applications/PyCharm.app/Contents/helpers/pydev/_pydev_bundle/pydev_umd.py", line 197, in runfile
    pydev_imports.execfile(filename, global_vars, local_vars)  # execute the script
  File "/Applications/PyCharm.app/Contents/helpers/pydev/_pydev_imps/_pydev_execfile.py", line 18, in execfile
    exec(compile(contents+"\n", file, 'exec'), glob, loc)
  File "/Users/user/sci/cgae/otp_equi.py", line 223, in <module>
    main()
  File "/Users/user/sci/cgae/otp_equi.py", line 215, in main
    results = execute(args)
  File "/Users/user/sci/cgae/otp_equi.py", line 114, in execute
    pred_sph = decoder(cg_features, cg_xyz)
  File "/Users/user/miniconda3/lib/python3.7/site-packages/torch/nn/modules/module.py", line 493, in __call__
    result = self.forward(*input, **kwargs)
  File "/Users/user/sci/cgae/cgae/equi.py", line 89, in forward
    output = layer(output.div(geometry.size(1) ** 0.5), geometry)  # Normalization of layers by number of atoms.
  File "/Users/user/miniconda3/lib/python3.7/site-packages/torch/nn/modules/module.py", line 493, in __call__
    result = self.forward(*input, **kwargs)
  File "/Users/user/sci/se3cnn/se3cnn/non_linearities/gated_block.py", line 41, in forward
    features = self.op(*args, **kwargs)  # [..., channel, ...]
  File "/Users/user/miniconda3/lib/python3.7/site-packages/torch/nn/modules/module.py", line 493, in __call__
    result = self.forward(*input, **kwargs)
  File "/Users/user/sci/se3cnn/se3cnn/point/operations.py", line 19, in forward
    k = self.kernel(rb - ra)  # [batch, a, b, i, j]
  File "/Users/user/miniconda3/lib/python3.7/site-packages/torch/nn/modules/module.py", line 493, in __call__
    result = self.forward(*input, **kwargs)
  File "/Users/user/sci/se3cnn/se3cnn/point/kernel.py", line 120, in forward
    Ys = self.sh(self.set_of_l_filters, r)  # [l_filter * m_filter, batch]
  File "/Users/user/sci/se3cnn/se3cnn/SO3.py", line 304, in spherical_harmonics_xyz
    out = spherical_harmonics(order, alpha, beta)  # [m, ...]
  File "/Users/user/sci/se3cnn/se3cnn/SO3.py", line 274, in spherical_harmonics
    Y = sh(Js, Ms, math.pi - beta.cpu().numpy(), alpha.cpu().numpy())
RuntimeError: Can't call numpy() on Variable that requires grad. Use var.detach().numpy() instead.
```

**Comment by bkmi** [comment]
> for me it looks like `examples/point/equivariance.py` should be put into the test files and use the same format as the other files in this directory.

I though the purpose of this file was so that a user would have a minimum example to learn how to test equivariance. I'm happy to put it in test but I think there must be some kind of note in the readme letting the user know about it. I would never read the tests as a user. I also think they serve slightly different purposes since the `examples/point/equivariance.py` code is commented.

In fact, a lot of this information is more-or-less in the tests already. I noticed a normal GatedBlock test is missing, but the Kernel is tested for rotation and parity already.

**Comment by mariogeiger** [comment]
Ok I agree it makes sense

**Pull request merged**